File: Sources/SnipKeyCore/TriggerContextAnalyzer.swift
import Foundation

public enum TriggerContextAnalyzer {
    public struct CompletedTrigger: Equatable {
        public let trigger: String
        public let deletionCount: Int

        public init(trigger: String, deletionCount: Int) {
            self.trigger = trigger
            self.deletionCount = deletionCount
        }
    }

    public static func activeQuery(in textBeforeCursor: String, triggerPrefix: Character = "#") -> String? {
        let characters = Array(textBeforeCursor)

        if characters.last == triggerPrefix {
            return ""
        }

        guard let endIndex = lastTriggerCharacterIndex(in: characters) else {
            return nil
        }

        var startIndex = endIndex
        while startIndex > 0, isTriggerCharacter(characters[startIndex - 1]) {
            startIndex -= 1
        }

        guard startIndex > 0, characters[startIndex - 1] == triggerPrefix else {
            return nil
        }

        return String(characters[startIndex...endIndex])
    }

    public static func completedTrigger(in textBeforeCursor: String, triggerPrefix: Character = "#") -> CompletedTrigger? {
        let characters = Array(textBeforeCursor)
        guard let trailingCharacter = characters.last, isTriggerTerminator(trailingCharacter) else {
            return nil
        }

        let committedText = String(characters.dropLast())
        guard let query = activeQuery(in: committedText, triggerPrefix: triggerPrefix), !query.isEmpty else {
            return nil
        }

        return CompletedTrigger(trigger: query, deletionCount: query.count + 2)
    }

    private static func lastTriggerCharacterIndex(in characters: [Character]) -> Int? {
        guard let lastIndex = characters.indices.last, isTriggerCharacter(characters[lastIndex]) else {
            return nil
        }

        return lastIndex
    }

    private static func isTriggerCharacter(_ character: Character) -> Bool {
        SnippetTriggerRules.isAllowedCharacter(character)
    }

    private static func isTriggerTerminator(_ character: Character) -> Bool {
        !isTriggerCharacter(character) && character != "#"
    }
}

File: Windows/SnipKey.Windows/Core/TriggerContextAnalyzer.cs
namespace SnipKey.WinApp.Core;

public static class TriggerContextAnalyzer
{
    public readonly record struct CompletedTrigger(string Trigger, int DeletionCount);

    public static string? ActiveQuery(string textBeforeCursor, char triggerPrefix = '#')
    {
        if (string.IsNullOrEmpty(textBeforeCursor))
        {
            return null;
        }

        if (textBeforeCursor[^1] == triggerPrefix)
        {
            return string.Empty;
        }

        var endIndex = textBeforeCursor.Length - 1;
        if (!SnippetTriggerRules.IsAllowedCharacter(textBeforeCursor[endIndex]))
        {
            return null;
        }

        var startIndex = endIndex;
        while (startIndex > 0 && SnippetTriggerRules.IsAllowedCharacter(textBeforeCursor[startIndex - 1]))
        {
            startIndex -= 1;
        }

        if (startIndex == 0 || textBeforeCursor[startIndex - 1] != triggerPrefix)
        {
            return null;
        }

        return textBeforeCursor[startIndex..(endIndex + 1)];
    }

    public static CompletedTrigger? CompletedTriggerIn(string textBeforeCursor, char triggerPrefix = '#')
    {
        if (string.IsNullOrEmpty(textBeforeCursor))
        {
            return null;
        }

        var trailingCharacter = textBeforeCursor[^1];
        if (!IsTriggerTerminator(trailingCharacter, triggerPrefix))
        {
            return null;
        }

        var committedText = textBeforeCursor[..^1];
        var query = ActiveQuery(committedText, triggerPrefix);
        if (string.IsNullOrEmpty(query))
        {
            return null;
        }

        return new CompletedTrigger(query, query.Length + 2);
    }

    private static bool IsTriggerTerminator(char character, char triggerPrefix)
    {
        return !SnippetTriggerRules.IsAllowedCharacter(character) && character != triggerPrefix;
    }
}